src/Http/Routing/src/Constraints/FileNameRouteConstraint.cs

```diff
@@ -0,0 +1,145 @@
+    ///     <listheader>  
+    ///         <term>Examples of route values that will be matched as file names</term>  
+    ///         <description>description</description>  
+    ///     </listheader>  
+    ///     <item>  
+    ///         <term><c>/a/b/c.txt</c></term>  
+    ///         <description>Final segment contains a <c>.</c> surrounded by other characters.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/hello.world.txt</c></term>  
+    ///         <description>Final segment contains a <c>.</c> surrounded by other characters.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>hello.world.txt</c></term>  
+    ///         <description>Final segment contains a <c>.</c> surrounded by other characters.</description>  
+    ///     </item> 
+    /// </list>
+    /// <list type="bullet">  
+    ///     <listheader>  
+    ///         <term>Examples of route values that will be rejected as non-file-names</term>  
+    ///         <description>description</description>  
+    ///     </listheader>  
+    ///     <item>  
+    ///         <term><c>/a/b/c</c></term>  
+    ///         <description>Final segment does not contain a <c>.</c>.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/a/b.d/c</c></term>  
+    ///         <description>Final segment does not contain a <c>.</c>.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/a/b.d/c/</c></term>  
+    ///         <description>Final segment is empty.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c>/a/b/.c</c></term>  
+    ///         <description>Final segment contains a <c>.</c> but is not surrounded by other characters.</description>  
+    ///     </item>
+    ///     <item>  
+    ///         <term><c></c></term>  
+    ///         <description>Value is empty</description>  
+    ///     </item>
+    /// </list>  
+    /// </para>
+    /// </remarks>
+    public class FileNameRouteConstraint : IRouteConstraint
+    {
+        /// <inheritdoc />
+        public bool Match(
+            HttpContext httpContext,
+            IRouter route,
+            string routeKey,
+            RouteValueDictionary values,
+            RouteDirection routeDirection)
+        {
+            if (routeKey == null)
+            {
+                throw new ArgumentNullException(nameof(routeKey));
+            }
+
+            if (values == null)
+            {
+                throw new ArgumentNullException(nameof(values));
+            }
+
+            if (values.TryGetValue(routeKey, out var obj) && obj != null)
+            {
+                var value = Convert.ToString(obj, CultureInfo.InvariantCulture);
+                return IsFileName(value);
+            }
+
+            // No value or null value.
+            return false;
+        }
+
+        internal static bool IsFileName(ReadOnlySpan<char> value)
```

Comment:
Certain characters are invalid in a file name, e.g. `:`, `&`. What about checking for those? To complicate things, invalid file name characters are probably different between operating systems.

Whether this constraint does or doesn't check for them then it would be worth commenting on in the XML docs.

Reply:
hmm, I thought I did document that. I'll double check. 

It's definitely intended that we don't validate these. 1 - it's platform specific, 2 - it's not aligned with the purpose of this